### ISSUE TITLE
Avoid manual downloading of dom0 rpm during bootstrapping

### ISSUE THREAD

**Issue by legoktm** [freedomofpress/securedrop-workstation#945]
I want to split out the first step from #942 as its own separate issue:
>  The rough flow is as follows:
> 1. configure a network-attached VM to download the dom0 rpm, copy and install it

https://workstation.securedrop.org/en/stable/admin/install.html#download-and-install-securedrop-workstation describes a lot of steps related to verifying the key, manually writing a .repo file, verifying and moving the RPM around and then finally installing it.

My proposed solution would be to have a `securedrop-keyring` RPM that is shipped in the Qubes repositories, so you can simply run `sudo dnf install securedrop-keyring` from dom0 and it provisions our public signing key in `/etc/pki/rpm-gpg/` (or elsewhere). Then the user creates the .repo file, and can run `sudo dnf install securedrop-workstation-dom0-config` and we're all set. Instead of directly trusting the SD release key, we would be trusting that the Qubes-signed package provided us with the correct key (certainly if Qubes is compromised, the whole workstation is compromised anyways).

This would largely require the Qubes team being willing to ship our key, figuring out details around update cadence, timelines, etc.

**Comment by rocodes** [comment]
@legoktm I think this is a good idea whether or not we end up working towards an ISO, although I think it's currently semi blocked on the last sentence you mentioned (bit more of an internal conversation about update cadence/making sure our upstream ask isn't too onerous).

What I'm wondering if we do this is why we wouldn't also ship the .repo file and have it be autoconfigured in the postinnst. I think we can find some way of visually confirming to the user before they proceed with the installation, and/or we can still support the legacy method of checking and importing the key manually for users who want to go the extra mile.

**Comment by rocodes** [comment]
Bumping this for a couple reasons, besides the initial benefits discussed above:

- https://github.com/freedomofpress/securedrop-workstation/issues/953 (/earlier https://github.com/freedomofpress/securedrop-workstation/issues/423) would be well-served by a package postinst. We should resolve this before our next key expiry this summer
- rather than configuring .repo files dynamically with Salt, it would be nice to be able to define a static .repo file(s). It would also lend more confidence about system state this way (eg: verifying that `securedrop-yum-keyring` and not `securedrop-yum-keyring-test` package is installed, vs manually searching for all the places the key could be)
- we might be able to automate some manual aspects of QA (such as switching between yum and yum-qa), or take advantage of some of DNF's built-in abilities (for example, an orchestration package that depends on a keyring package and reprovisions on keyring package changes if a user flips between prod and test key - but now I'm spitballing a little bit)

**Comment by lsd-cat** [comment]
Just flagging that depending on the Qubes team preferences and choices, we could also ask to ship our repository by default but keep it disabled, and then asks users to just issue

`
dnf config-manager --set-enabled securedrop-something
`

Or ship the repo file as a package too as it seems they are doing here https://www.qubes-os.org/doc/installing-contributed-packages/

**Comment by rocodes** [comment]
After our conversation with the Qubes team, it sounds like they're open to having a small rpm with our key and repo file in qubes-contrib.

I made a small repo with an rpm .spec file, the .repo file, the pubkey, and our containerized build stuff (I don't know that Qubes will use the latter, but tbd); the spec file was as below:

```
Name:		securedrop-workstation-keyring
Version:	0.1.0
Release:	1%{?dist}
Summary:	SecureDrop Workstation Keyring

# For reproducible builds:
#
#   * Ensure that SOURCE_DATE_EPOCH env is honored and inherited from the
#     last changelog entry, and enforced for package content mtimes
%define source_date_epoch_from_changelog 1
%define use_source_date_epoch_as_buildtime 1
%define clamp_mtime_to_source_date_epoch 1
#   * By default, changelog entries for the last two years of the current time
#     (_not_ SOURCE_DATE_EPOCH) are included, everything else is discarded.
#     For easy reproducibility we'll keep everything
%define _changelog_trimtime 0
%define _changelog_trimage 0
#   * _buildhost varies based on environment, we build with containers but
#     ensure this is the same regardless
%global _buildhost %{name}
#   * optflags is for multi-arch support: otherwise rpmbuild sets 'OPTFLAGS: -O2 -g -march=i386 -mtune=i686'
%global optflags -O2 -g
# To ensure forward-compatibility of RPMs regardless of updates to the system
# Python, we disable the creation of bytecode at build time via the build
# root policy.
%undefine py_auto_byte_compile

License:	AGPLv3
# TODO
URL:		https://github.com/rocodes/securedrop-workstation-keyring
Source:		%{url}/archive/refs/tags/%{version}.tar.gz#/%{name}-%{version}.tar.gz
# See: https://docs.fedoraproject.org/en-US/packaging-guidelines/SourceURL/#_troublesome_urls

BuildArch:		noarch
BuildRequires:	systemd-rpm-macros

%description
This package contains the SecureDrop Release public key and yum .repo file
used to bootstrap installation of SecureDrop Workstation.

%prep
%setup -q -n files

%build
# No building necessary

%install
install -m 755 -d %{buildroot}/etc/yum.repos.d
install -m 755 -d %{buildroot}/etc/pki/rpm-gpg
install -m 644 %{_builddir}/files/securedrop-workstation-dom0.repo %{buildroot}/etc/yum.repos.d/
install -m 644 %{_builddir}/files/securedrop-release-signing-pubkey-2021.asc %{buildroot}/etc/pki/rpm-gpg/RPM-GPG-KEY-securedrop-workstation

%files
/etc/pki/rpm-gpg/RPM-GPG-KEY-securedrop-workstation
/etc/yum.repos.d/securedrop-workstation-dom0.repo

%preun
# see https://github.com/rpm-software-management/rpm/issues/2577
# naive idea - this does not work
key_id=$(rpm -q gpg-pubkey --qf '%{NAME}-%{VERSION}-%{RELEASE}\t%{SUMMARY}\n' | grep SecureDrop | cut -f1 -d' ')
rpm -e $key_id || true

%post
# naive idea - this does not work
sudo rpm --import /etc/pki/rpm-gpg/RPM-GPG-KEY-securedrop-workstation

%changelog
# TODO
```

but there's a hiccup, as I've indicated in comments - we can't use `rpm` commands in the rpm %post or %preun even if they're for bootstrapping other keys (our key), because rpm maintains a transaction lock over the rpm database while its installing and updating packages. 

afaict, the best we can do re key management is provide a bootstrapping script that the user runs after the initial RPM is downloaded and the transaction lock on the rpm database is lifted.

**Comment by deeplow** [comment]
Per discussion in a meeting today, we'll be postponing the final QA for monday as by then we'll probably already have the final instructions for https://github.com/freedomofpress/securedrop-workstation-docs/pull/338/.

In the meantime we'll prepare the QA artifacts so we have everything ready for the next stage of QA.

**Comment by deeplow** [comment]
Solved with the release of 1.5.0 🥳
